interop-cats/jvm/src/main/scala/scalaz/zio/interop/catszmanagedjvm.scala

```diff
@@ -0,0 +1,109 @@
+package scalaz.zio.interop
+
+import cats.effect.Resource.{ Allocate, Bind, Suspend }
+import cats.effect.{ ExitCase, LiftIO, Resource, Sync, IO => CIO }
+import cats.{ effect, Bifunctor, Monad, MonadError, Monoid, Semigroup, SemigroupK }
+import scalaz.zio.{ Reservation, ZIO, ZManaged }
+import scalaz.zio.Runtime
+
+trait CatsZManagedSyntax {
+
+  implicit class ResourceSyntax[A](resource: Resource[CIO, A]) {
+    def toManaged[R, E](
+      implicit l: LiftIO[ZIO[R, Throwable, ?]]
+    ): ZManaged[R, Throwable, A] = {
+      def convert[A1](resource: Resource[CIO, A1]): ZManaged[R, Throwable, A1] =
+        resource match {
+          case Allocate(res) =>
+            ZManaged.unwrap(
```

Comment:
I think we actually have a limitation with `ZManaged` currently as it does not allow finalizers to inspect the exit condition. Without fixing that, we can't faithfully translate the finalizer inside `Allocated`.

Reply:
I've updated the implementation. This should now be a faithful conversion from resource

Reply:
Ok, that's quite clever, but I think it still won't work correctly. The problem is that the correct finalizer (as determined by the `ExitCase`) will only be executed when the resulting `ZManaged` is used directly.

When you compose it with others using `flatMap`, I think the finalizer won't run correctly (e.g. it'll run right after executing the `reserve`, which is not right). Moreover, the finalizer stored in the `Reservation` won't be the right one: it's constructed by `r(ExitCase.Completed)`, which is not always true.

A simpler definition would be something like this:
```scala
// ZManaged.scala
object ZManaged {
  def makeExit(acquire: ZIO[R, E, A])(release: (A, Exit[E, _]) => ZIO[R, Nothing, _])): ZManaged[R, E, A] = ???
}

// catszmanagedjvm.scala
case Allocate(res) => 
  ZManaged.make(l.liftIO(res)) { case (exit, (_, r) =>
    l.liftIO(r(exitToExitCase(exit))).orDie
  }.map(_._1)
```

And we'd also need a function `def exitToExitCase[E](exit: Exit[E, _]): ExitCase[E]`.

Reply:
This is inspired by how it's done in cats-effects' allocated https://github.com/typelevel/cats-effect/blob/94a7498952c4b78c4cdbdf6d609c347e26c85a91/core/shared/src/main/scala/cats/effect/Resource.scala#L188-L227.

The ExitCase can be one of {Completed, Canceled, Error}. Like this we would push the handlers 
for { Canceled, Error } into the reservation. Once the resource is reserved it is already allocated here and the only case we have to deal with is cleaning up after a completed allocation.

The only finalizers that should run directly after reserve should be {Canceled and Failed} which I *think* is fine.

BTW. are you aware where the test suites / laws / etc. for Resource are in cats effect? I failed to find them when I looked

Reply:
Ahh I see. Thanks, that makes sense and I understand the logic now. 

Would be good if you could add a specific test for a `ZManaged` converted from a `Resource` used with `ZManaged#flatMap` with other `ZManaged` instances.

The `Resource` tests are here: https://github.com/typelevel/cats-effect/blob/master/laws/shared/src/test/scala/cats/effect/ResourceTests.scala

Reply:
Thank you. I'll add tests for this and also for the zmanaged -> resource conversion